### PR TITLE
Feature/hyphenation Added hyphenation

### PR DESCRIPTION
**Pull request by yashha**
TODO: adding a test would be good.

We have to switch to a higher supported node version for this update.
Version 10.22 is not working. It has to be version 11.0 or higher.

EDIT: Added Tests

### CODE REVIEW

**Comment by danburzo** [comment]
@yashha thank you for this, the feature sounds great! 

My first impression is the hyphenopoly requirement for Node 11 is accidental, since [`TextDecoder`](https://nodejs.org/api/util.html#util_class_util_textdecoder) seems to be available as `require('util').TextDecoder` since Node 8.3. If that's the only incompatibility with Node 10.x, I'd rather work with the author towards a fix than raise percollate's requirements. 

I'll have more detailed feedback after a closer look.

**Comment by danburzo** [comment]
I had the chance to take a closer look at the PR, this is looking very nice! I have a few thoughts on the implementation, which I hope make more sense as a single message here than inline comments in the code. 

Extracting the document language is, I think, something that is useful to have in the HTML template as well, so I propose bringing it into `index.js` and adding it as the second argument to `hyphenateHtml`:

```diff
diff --git a/index.js b/index.js
index 0ba5778..18c261c 100755
--- a/index.js
+++ b/index.js
@@ -21,6 +21,7 @@ const humanDate = require('./src/util/human-date');
 const outputPath = require('./src/util/output-path');
 const addExif = require('./src/exif');
 const { hyphenateHtml } = require('./src/hyphenate');
+const { textToLang } = require('./src/util/language');
 
 const {
 	ampToHtml,
@@ -238,8 +239,8 @@ async function cleanup(url, options) {
 
 		const parsed = R.parse() || {};
 
-		// Hyphenate the text
-		parsed.content = hyphenateHtml(parsed.content);
+		const textContent = sanitizer.sanitize(parsed.textContent); 
+		const lang = textToLang(textContent);
 
 		out.write(' ✓\n');
 
@@ -253,9 +254,13 @@ async function cleanup(url, options) {
 				sanitizer.sanitize(parsed.excerpt, { RETURN_DOM: true })
 			),
 			content: serializer(
-				sanitizer.sanitize(parsed.content, { RETURN_DOM: true })
+				hyphenateHtml(
+					sanitizer.sanitize(parsed.content, { RETURN_DOM: true }),
+					lang
+				)
 			),
-			textContent: sanitizer.sanitize(parsed.textContent),
+			textContent,
+			lang,
 			length: parsed.length,
 			siteName: sanitizer.sanitize(parsed.siteName),
 			remoteResources
```

...and then placing it in the HTML:

```diff --git a/templates/default.html b/templates/default.html
index 74cfcfe..c9a6f40 100644
--- a/templates/default.html
+++ b/templates/default.html
@@ -38,7 +38,7 @@
 			</ol>
 		</nav>
 		{% endif %} {% for item in items %}
-		<article id="{{ item.id }}" class="article">
+		<article id="{{ item.id }}" class="article" lang="{{item.lang}}">
 			<header class="article__header">
 				<h1 class="article__title">
 					{{ item.title }}
```

Speaking of `hyphenateHtml`, changing the first argument from a HTML _string_ to a DOM element has some advantages:

* We don't have to parse and stringify the HTML just to hyphenate;
* The DOM traversal code can become much simpler and more robust.

```js
function hyphenateHtml(el, lang) {
	const hyphenate = getHypenatorByLang(lang);
	const it = el.ownerDocument.createNodeIterator(el, 4);
	let node;
	while (node = it.nextNode()) {
		node.textContent = hyphenate(node.textContent);
	}
	return el;
}
```

(notice `lang` is now the second parameter to the function).

Finally, while the hyphenation seems to work pretty well and it makes sense to have on by default, it would be useful to figure out a mechanism for users to bail out of it.

**Comment by danburzo** [comment]
One more thing: the browser version of Hyphenopoly, which does DOM-based hyphenation, might contain more useful things we should be doing on the Node.js side as well. I'll dig in and let you know if I find anything!

File: src/util/language.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+const franc = require('franc-all');
+const iso6393 = require('iso-639-3');
```
Comment (danburzo) [inline src/util/language.js:2]: We can make this more efficient by using the map from `require('iso-639-3/to-1')` to find the equivalent ISO 639-1 code from the ISO 639-3.

**Comment by yashha** [comment]
> Speaking of hyphenateHtml, changing the first argument from a HTML string to a DOM element has some advantages:

Hm you get from Readablity only html, you will always have to parse and stringify it.

Thanks for the comments.

**Comment by danburzo** [comment]
That part could have used more details, sorry about that. 

Right now the flow for PDF / HTML is the following:

1. Readability takes a DOM and produces `content` as stringified HTML, and `textContent` as plain-text
2. The *sanitizer* (DOMPurify) takes the stringified HTML, parses & cleans it, and produces a DOM element because of the `{ RETURN_DOM: true }` option.
3. The DOM element is then turned back to a HTML string by the *serializer*.

What I was proposing is introducing the hyphenation step on the DOM element between steps 2 and 3, to avoid parsing & stringifying again after step 3 just to perform the hyphenation.

**Comment by danburzo** [comment]
Okay, I can confirm that the serializer (DOMPurify) has the ability to work with a DOM element instead of a HTML string, therefore in #115 I make Readability emit a DOM element in `parsed.content`, and only serialize to HTML / XHTML as the last step. 

After I merge it in, we should be able to add the hyphenation as:

```diff --git a/index.js b/index.js
index b4ef24a..b6c96cc 100755
--- a/index.js
+++ b/index.js
@@ -264,7 +264,10 @@ async function cleanup(url, options) {
 				sanitizer.sanitize(parsed.excerpt, { RETURN_DOM: true })
 			),
 			content: serializer(
-				sanitizer.sanitize(parsed.content, { RETURN_DOM: true })
+				hyphenateHtml(
+					sanitizer.sanitize(parsed.content, { RETURN_DOM: true }),
+					lang
+				)
 			),
 			textContent: sanitizer.sanitize(parsed.textContent),
 			length: parsed.length,
```

(with `hyphenateHtml` made to work on a DOM element)

How does that sound? Hopefully merging #115 into master doesn't give you a headache with conflicts.

**Comment by danburzo** [comment]
The changes look good! Unfortunately I have some personal issues to attend to, but I hope to be able to get back to this soon.

**Comment by danburzo** [comment]
Hyphenation

I had the chance to sit down with the code and I think we are mostly there! Here are a few more ideas:

### Language detection

The `lang` attribute, present on the `html` tag on [a surprising number of pages](https://almanac.httparchive.org/en/2020/markup#document-language) should help us short-circuit the language detection code. 

(An ideal solution would also take into account the element-level `lang` attribute which, even though the Web Almanac did not look at specifically, I suspect is only used on very few pages; therefore it would be overkill to handle it in this first iteration.)

### The `--hyphenate, --no-hyphenate` options

Manual hyphenation relies on the insertion of soft hyphens, which have the unfortunate side-effect of carrying over when pieces of text are copied to the clipboard. I propose:

* Have `--hyphenate` and `--no-hyphenate` flags to explicitly enable/disable hyphenation
* Enable hyphenation by default for PDFs — because PDFs are not reflowable (they have a fixed layout), but also because it seems soft hypens are _not_ present in the text copied to the clipboard.
* Keep hyphenation disabled by default for EPUB and HTML. These are two formats where people are more likely to highlight/copy text, which could be bothersome. Also, some EPUB readers have their own, more sophisticated hyphenation, and their own options around hyphenation. 

### Default CSS

* `hyphens: auto` regardless of the hyphenation option allows the browsers to use their own hyphenators, where available, for EPUB and HTML
* `text-align: justify` — I have a strong preference against text justification, I would rather have users opt _into_ this behavior with custom CSS. 


Let me know what you think!

**Comment by yashha** [comment]
Sounds good to me! I will adapt the pull request.
I think for PDF the default CSS should be justified because it then looks more like a book and I like it, but that is just my preference. But we can make it as opt in when you want, I will set it to justified in the webservice though.

**Comment by danburzo** [comment]
@yashha let me know when I can merge this in! I'm not sure if the ball is in my court after your last batch of commits :-P

**Comment by yashha** [comment]
I think in pdf by default hyphenation and justify should be enabled.
When you only have hyphenation enabled you don't see an effect.
What do you think?
What should an "--hyphenate" option do?

**Comment by danburzo** [comment]
> I think in pdf by default hyphenation and justify should be enabled.
> When you only have hyphenation enabled you don't see an effect.
> What do you think?

The reason why I wasn't sure about this is, since Hyphenopoly (and, in general, the Liang-Knuth algorithm) errs on the side of fewer-but-correct hyphens, it makes justified text more likely to produce [undesired effects](https://en.wikipedia.org/wiki/River_(typography)), but I guess it doesn't hurt to experiment with justifying the text by default in PDFs. 

>  What should an "--hyphenate" option do?

Without any option provided we hyphenate PDFs. An explicit `--hyphenate` flag would enable hyphenation regardless of the file type; conversely, a `--no-hyphenate` will disable hyphenation regardless of the file type.

```js
let should_hyphenate = options.hyphenate === undefined ? (type === 'pdf') : options.hyphenate;
```

**Comment by yashha** [comment]
You can review now, whether it can be merged. :)

**Comment by yashha** [comment]
You can review now, whether it can be merged. :)

**Comment by danburzo** [comment]
Brilliant, thank you! I'll pull it in and make a couple of cosmetic changes, and I think we're good to go. 🎉

**Pull request merged**